pkg/session/bootstrap.go

```diff
@@ -3090,8 +3096,26 @@ func upgradeToVer212(s sessiontypes.Session, ver int64) {
 	if ver >= version212 {
 		return
 	}
+	// version212 changed a lots of runaway related table.
+	// 1. switchGroup: add column `switch_group_name` to `mysql.tidb_runaway_watch` and `mysql.tidb_runaway_watch_done`.
 	doReentrantDDL(s, "ALTER TABLE mysql.tidb_runaway_watch ADD COLUMN `switch_group_name` VARCHAR(32) DEFAULT '' AFTER `action`;", infoschema.ErrColumnExists)
 	doReentrantDDL(s, "ALTER TABLE mysql.tidb_runaway_watch_done ADD COLUMN `switch_group_name` VARCHAR(32) DEFAULT '' AFTER `action`;", infoschema.ErrColumnExists)
+	// 2. modify column `plan_digest` type, modify column `time` to `start_time,
+	// modify column `original_sql` to `sample_sql` and unique union key to `mysql.tidb_runaway_queries`.
+	// add column `sql_digest`.
+	doReentrantDDL(s, "ALTER TABLE mysql.tidb_runaway_queries ADD COLUMN `sql_digest` varchar(64) DEFAULT '' AFTER `original_sql`;", infoschema.ErrColumnExists)
+	// add column `repeats`.
+	doReentrantDDL(s, "ALTER TABLE mysql.tidb_runaway_queries ADD COLUMN `repeats` int DEFAULT 0 AFTER `time`;", infoschema.ErrColumnExists)
+	// modify column `time` to `start_time` and index(time).
+	doReentrantDDL(s, "ALTER TABLE mysql.tidb_runaway_queries CHANGE COLUMN `time` `start_time` TIMESTAMP NOT NULL", infoschema.ErrColumnExists)
+	doReentrantDDL(s, "ALTER TABLE mysql.tidb_runaway_queries DROP INDEX `time_index`", dbterror.ErrCantDropFieldOrKey)
+	doReentrantDDL(s, "ALTER TABLE mysql.tidb_runaway_queries ADD INDEX time_index(start_time) COMMENT 'accelerate the speed when querying with active watch'", dbterror.ErrDupKeyName)
+	// modify column `original_sql` to `sample_sql`.
+	doReentrantDDL(s, "ALTER TABLE mysql.tidb_runaway_queries CHANGE COLUMN `original_sql` `sample_sql` TEXT NOT NULL", infoschema.ErrColumnExists)
+	// modify column type of `plan_digest`.
+	doReentrantDDL(s, "ALTER TABLE mysql.tidb_runaway_queries MODIFY COLUMN `plan_digest` varchar(64) DEFAULT '';", infoschema.ErrColumnExists)
+	// add union key(resource_group_name, sql_digest, plan_digest, match_type)
+	doReentrantDDL(s, "ALTER TABLE mysql.tidb_runaway_queries ADD UNIQUE KEY runaway_task(resource_group_name, sql_digest, plan_digest, match_type);", dbterror.ErrDupKeyName)
```

Comment:
Why this key can be `unique`? What if 2 batch of record are flushed within 20s?

Reply:
> Why this key can be unique?

After discussing with the PM, we can use these params as unique cc @songrijie 

> What if 2 batch of record are flushed within 20s?

When refreshed within 20s, it will be updated by `ON DUPLICATE KEY UPDATE` which means the second batch will overlap previous's record's repeats.
https://github.com/pingcap/tidb/pull/55664/files#diff-d740f8f40d52887b1bee7292178ab85c36b151391a2afce9b58de3803e8d3af0R107
```bash
builder.WriteString(" ON DUPLICATE KEY UPDATE repeats = repeats + 1 + VALUES(repeats)")
```

Reply:
Ok. Got it.
Then, another question: if all the runaway records of the same query are flushed into the same row, how do you expect our user to use this row? E.g, how can I know how many new records are caught in the last 1 hour?

Reply:
After discussing with the PM, we will flush every 30 seconds and delete the union key, writing a new record(with compressing and repeats field) each time